File: src/data/skills.ts
import { existsSync, readdirSync, statSync, readFileSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';
import { execSync } from 'child_process';
import type { OcSkill, SkillScope } from '../types.js';

const OPENCLAW_ROOT = join(homedir(), '.openclaw');

function parseSkillMd(filePath: string): { name: string; description: string; raw: string } {
  const raw = readFileSync(filePath, 'utf-8');

  // Try YAML frontmatter (---\nname: ...\ndescription: ...\n---)
  const fmMatch = raw.match(/^---\s*\n([\s\S]*?)\n---/);
  if (fmMatch) {
    const block = fmMatch[1];
    const nameMatch = block.match(/^name:\s*(.+)$/m);
    const descMatch = block.match(/^description:\s*(.+)$/m);
    if (nameMatch || descMatch) {
      return {
        name: nameMatch?.[1]?.trim() ?? '',
        description: descMatch?.[1]?.trim() ?? '',
        raw,
      };
    }
  }

  // Fall back: first # heading = name, first non-empty non-heading paragraph = description
  const lines = raw.split('\n');
  let name = '';
  let description = '';
  let pastHeading = false;

  for (const line of lines) {
    const trimmed = line.trim();
    if (!name && trimmed.startsWith('# ')) {
      name = trimmed.replace(/^#+\s*/, '');
      pastHeading = true;
      continue;
    }
    if (pastHeading && !description && trimmed && !trimmed.startsWith('#')) {
      description = trimmed;
      break;
    }
  }

  return { name, description, raw };
}

function findBuiltInSkillsDir(): string | null {
  // Try npm root -g first
  try {
    const npmRoot = execSync('npm root -g 2>/dev/null', { encoding: 'utf-8' }).trim();
    if (npmRoot) {
      const p = join(npmRoot, 'openclaw', 'skills');
      if (existsSync(p)) return p;
    }
  } catch { /* fall through */ }

  // Fallback: known install locations
  const candidates = [
    join(homedir(), '.npm-global', 'lib', 'node_modules', 'openclaw', 'skills'),
    '/usr/local/lib/node_modules/openclaw/skills',
    '/usr/lib/node_modules/openclaw/skills',
  ];
  for (const p of candidates) {
    if (existsSync(p)) return p;
  }
  return null;
}

function loadSkillsFromDir(dir: string, scope: SkillScope): OcSkill[] {
  if (!existsSync(dir)) return [];
  const items: OcSkill[] = [];

  for (const entry of readdirSync(dir)) {
    const skillDir = join(dir, entry);
    try {
      if (!statSync(skillDir).isDirectory()) continue;
      const skillFile = join(skillDir, 'SKILL.md');
      if (!existsSync(skillFile)) continue;

      const { name, description, raw } = parseSkillMd(skillFile);
      items.push({
        kind: 'skill',
        id: `${scope}#${entry}`,
        name: name || entry,
        description,
        scope,
        filePath: skillFile,
        fullContent: raw,
      });
    } catch {
      // skip malformed entries
    }
  }

  return items.sort((a, b) => a.name.localeCompare(b.name));
}

export function loadSkills(): OcSkill[] {
  const results: OcSkill[] = [];

  // Built-in skills (shipped with npm package)
  const builtInDir = findBuiltInSkillsDir();
  if (builtInDir) {
    results.push(...loadSkillsFromDir(builtInDir, 'built-in'));
  }

  // User-installed skills
  results.push(...loadSkillsFromDir(join(OPENCLAW_ROOT, 'skills'), 'installed'));

  return results;
}

File: internal/data/skills.go
package data

import (
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var fmRe = regexp.MustCompile(`(?s)^---\s*\n(.*?)\n---`)
var fmNameRe = regexp.MustCompile(`(?m)^name:\s*(.+)$`)
var fmDescRe = regexp.MustCompile(`(?m)^description:\s*(.+)$`)

func parseSkillMd(filePath string) (name, description, raw string) {
	b, err := os.ReadFile(filePath)
	if err != nil {
		return "", "", ""
	}
	raw = string(b)

	if m := fmRe.FindStringSubmatch(raw); m != nil {
		block := m[1]
		if nm := fmNameRe.FindStringSubmatch(block); nm != nil {
			name = strings.TrimSpace(nm[1])
		}
		if dm := fmDescRe.FindStringSubmatch(block); dm != nil {
			description = strings.TrimSpace(dm[1])
		}
		if name != "" || description != "" {
			return
		}
	}

	pastHeading := false
	for _, line := range strings.Split(raw, "\n") {
		trimmed := strings.TrimSpace(line)
		if name == "" && strings.HasPrefix(trimmed, "# ") {
			name = strings.TrimLeft(trimmed, "# ")
			pastHeading = true
			continue
		}
		if pastHeading && description == "" && trimmed != "" && !strings.HasPrefix(trimmed, "#") {
			description = trimmed
			break
		}
	}
	return
}

func findBuiltInSkillsDir() string {
	if out, err := exec.Command("npm", "root", "-g").Output(); err == nil {
		p := filepath.Join(strings.TrimSpace(string(out)), "openclaw", "skills")
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}
	home, _ := os.UserHomeDir()
	candidates := []string{
		filepath.Join(home, ".npm-global", "lib", "node_modules", "openclaw", "skills"),
		"/usr/local/lib/node_modules/openclaw/skills",
		"/usr/lib/node_modules/openclaw/skills",
	}
	for _, p := range candidates {
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}
	return ""
}

func loadSkillsFromDir(dir, scope string) []OcSkill {
	if _, err := os.Stat(dir); err != nil {
		return nil
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	var items []OcSkill
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		skillFile := filepath.Join(dir, entry.Name(), "SKILL.md")
		if _, err := os.Stat(skillFile); err != nil {
			continue
		}
		name, desc, raw := parseSkillMd(skillFile)
		if name == "" {
			name = entry.Name()
		}
		items = append(items, OcSkill{
			ID:          scope + "#" + entry.Name(),
			Name_:       name,
			Description: desc,
			Scope:       scope,
			FilePath:    skillFile,
			FullContent: raw,
		})
	}
	sort.Slice(items, func(i, j int) bool { return items[i].Name_ < items[j].Name_ })
	return items
}

func LoadSkills() []OcSkill {
	var result []OcSkill
	if dir := findBuiltInSkillsDir(); dir != "" {
		result = append(result, loadSkillsFromDir(dir, "built-in")...)
	}
	home, _ := os.UserHomeDir()
	result = append(result, loadSkillsFromDir(filepath.Join(home, ".openclaw", "skills"), "installed")...)
	return result
}
